### PR TITLE
proximity notification for map cache and waypoint popups (fix #417)

### LINKED ISSUE #417: Notification sound while seeking
It would be nice to have a notification sound, while seeking a cache and the distance to the cache goes under 50 or 40 meters.
This is also avaiable in garmin devices.

### PR DESCRIPTION
**Pull request by moving-bits**
Adds a notification sound when approaching the cache or waypoint currently show in the map popup.
There are two distances configurable: One for single beep, one for double beep.

add distance beep feature to map cache and waypoint popups (AbstractDialogFragmentWithBeep)
add configuration settings, strings and preference keys
add AbstractSeekbarPreference
refactor existing BroutherThresholdPreference and WpThresholdPreference to use it

### CODE REVIEW

**Comment by SammysHP** [comment]
AFAICS this does not add the feature to the compass activity, correct? I think the compass activity would be the more important location for most users interested in this feature.

**Comment by moving-bits** [comment]
> AFAICS this does not add the feature to the compass activity, correct? I think the compass activity would be the more important location for most users interested in this feature.

Yes, compass activity is unchanged for now, as the cache popup in map activities is the most common use case for me, followed by the waypoint popup. But compass activity may be added later (probably with separate PR).

**Comment by Lineflyer** [comment]
Interesting!!
I am in favor to give it a try if considered OK codewise by others of the team.

**Comment by okainov** [comment]
Looking at #417 I have couple of comments:

Can you describe a use case (-s)? There were a lot of questions from dev team before exactly about use cases, and when I try to think of it, I realize that it's really somehow questionable.

- If you're just wandering around (walking\driving\cycling) with your audio on then probably you expect notifications when *any* cache is nearby. This PR doesn't solve this.
- If you're in the forest looking for the final cache box, they I'd really think that you either navigate yourself by simply looking at the map or using dedicated "compass" activity. Navigating when popup is open is not really the case - you don't see half of the map. This PR doesn't solve this.
- ..but when you're using compass, I assume, you have the phone in your hands and on, so you see the screen and not really blindly following the theoretical beeps. I can hardly imagine that someone walks in the forest to the cache without a screen.
- Just a side note why I'm saying "forest" - because in the cities compass doesn't help due to some GPS lags or bad precision, it's much more efficient to navigate yourself using the map (because you know where you are usually, even without GPS)


> I'm meanwhile in favor of generating speach output like "270 meters, 8 o'clock,...". 

Any comment, why have you decided to go with just beeps? There were a lot of pros in the comments for more detailed sound. Also it makes even more sense when you don't see a compass.

And for sure wording in settings need to be improved, "1" and "2" are not really helping :)

Also do you remember my comment about extracting log-scale-based slider to some reusable "base"? Back then you said WP slider it's the only log-based slider in the app. But now you've got two more. I still think that you should extract the logic to some "AbstractLogBasedSlider" or "ILogSlider" ;)

**Comment by moving-bits** [comment]
> Also do you remember my comment about extracting log-scale-based slider to some reusable "base"? Back then you said WP slider it's the only log-based slider in the app. But now you've got two more. I still think that you should extract the logic to some "AbstractLogBasedSlider" or "ILogSlider" ;)

WP slider is linear. "OC max distance" would have become the only log based slider then, but we decided not to merge the "OC max distance" function.

Now we have two log based sliders for the proximity settings, and I refactored the seekbar preference to `AbstractSeekbarPreference` (which is linear, but extendable) and `AbstractDistanceBeepPreference` (which extends it to be log based).

**Comment by moving-bits** [comment]
> Can you describe a use case (-s)? There were a lot of questions from dev team before exactly about use cases, and when I try to think of it, I realize that it's really somehow questionable.

My use case is quite simple: I'm on a geo trail, cycling from cache to cache of this trail, using the map display. When I'm done with a certain cache, I tap on the next cache on the map and continue.
The same works if you're looking for waypoints during a multi stage cache.

I do understand that there are probably tons of other use cases for a proximity beep, but before never ever getting a proximity beep implemented in c:geo at all I started with the ones most pressing to me :-) (no offense to anyone intended)

File: main/res/values/strings.xml [hunk 1]
```diff
@@ -596,6 +596,8 @@
     <string name="init_showwaypoint_description">If less than the given amount of caches are displayed on the map, their waypoints are shown additionally.</string>
     <string name="init_brouterThreshold">Calculate route</string>
     <string name="init_brouterThreshold_description">Maximum distance up to which brouter helper app (if installed) will calculate a route.</string>
+    <string name="init_distanceBeep">Distance beep</string>
```
Comment (bekuno) [inline main/res/values/strings.xml]: I suggest not to use the word beep.
What about "Distance signaling"?

File: main/res/values/strings.xml [hunk 2]
```diff
@@ -596,6 +596,8 @@
...
+    <string name="init_summary_distanceBeep">Play one or two beeps when coming close to the active cache.</string>
```
Comment (bekuno) [inline main/res/values/strings.xml]: And here:
Generates one or two signaling tones when you are approaching the enabled cache.

**Comment by SammysHP** [comment]
As explained in the issue, I think this goes back to the implementation in Garmin GPS receivers. If you navigate to a POI the device beeps a single time when you are close to the selected POI (depending on speed 10-100 m before).

My use case is simple: select a cache, start compass, turn off the display, put phone in pocket and go into the direction of the cache (e.g. along a path or towards a landmark). When it beeps, stop, get the phone and then use the compass to finish the last meters.

**Comment by moving-bits** [comment]
> My use case is simple: select a cache, start compass, turn off the display, put phone in pocket and go into the direction of the cache (e.g. along a path or towards a landmark). When it beeps, stop, get the phone and then use the compass to finish the last meters.

That sounds reasonable and pretty similar to my use case. I'll look into it and see if I can add this as well.

File: main/res/values/strings.xml (hunk 1 above)
Comment (moving-bits) [reply main/res/values/strings.xml]: texts for settings title and info texts adjusted

File: main/res/values/strings.xml (hunk 2 above)
Comment (moving-bits) [reply main/res/values/strings.xml]: texts for settings title and info texts adjusted

**Comment by okainov** [comment]
>cycling <...>. When I'm done with a certain cache, I tap on the next cache on the map and continue.

Being on a bike, how do you understand where to go? That's what  I meant - two types of beep do not help you to understand that you're getting closer. Dynamic frequency would be more suitable (think of parking sensors or Geiger counter)

**Comment by moving-bits** [comment]
> My use case is simple: select a cache, start compass, turn off the display, put phone in pocket and go into the direction of the cache (e.g. along a path or towards a landmark). When it beeps, stop, get the phone and then use the compass to finish the last meters.

added proximity notification to compass activity with 98971ad

**Comment by bekuno** [comment]
> Being on a bike, how do you understand where to go? That's what I meant - two types of beep do not help you to understand that you're getting closer. Dynamic frequency would be more suitable (think of parking sensors or Geiger counter)

I think that this will be a more complex solution, for a more special use case.

**Comment by SammysHP** [comment]
> Being on a bike, how do you understand where to go? That's what I meant - two types of beep do not help you to understand that you're getting closer. Dynamic frequency would be more suitable (think of parking sensors or Geiger counter)

This should be covered by the TTS output. It reads the direction and distance and is available in the compass activity.

**Comment by Lineflyer** [comment]
> >My use case is simple: select a cache, start compass, turn off the display, put phone in pocket and go into the direction of the cache (e.g. along a path or towards a landmark). When it beeps, stop, get the phone and then use the compass to finish the last meters.
>
> That sounds reasonable and pretty similar to my use case. I'll look into it and see if I can add this as well.

Indeed, that would also be my preferred use case.

**Comment by Lineflyer** [comment]
@moving-bits 
Regarding this use case:
> My use case is simple: select a cache, start compass, turn off the display, put phone in pocket and go into the direction of the cache (e.g. along a path or towards a landmark). When it beeps, stop, get the phone and then use the compass to finish the last meters.

I have seen you implemented this as well for the compass?
Did you check whether your new feature is also affected by #7057 and maybe we could -in case - fix that problem as well?

**Comment by moving-bits** [comment]
> I have seen you implemented this as well for the compass?

Yes

> Did you check whether your new feature is also affected by #7057 and maybe we could -in case - fix that problem as well?

Haven't tested for this yet, but AFAIR we (or Android?) switch off GPS when screen is turned off, which means, that we don't receive any coord updates, which in turn means no proximity notifications, neither in map popups nor in compass mode. This PR does not change this behavior.

**Comment by moving-bits** [comment]
> Haven't tested for this yet, but AFAIR we (or Android?) switch off GPS when screen is turned off, which means, that we don't receive any coord updates [...]

I haven't tested any of the following apps, but according to their description on Google Play they allow to stay GPS connected even if screen is turned off:
- https://play.google.com/store/apps/details?id=com.silentlexx.gpslock
- https://play.google.com/store/apps/details?id=org.bruxo.gpsconnected
- https://play.google.com/store/apps/details?id=com.webmajstr.gpson

(and many others)

Maybe this solves the problem mentioned in #7057?

**Comment by Lineflyer** [comment]
The problem with #7057 is, that it is a regression. It used to work but stopped working at some point. Might be due to targetsdk change or simply due to power management on newer androud devices. I have not tested it on different versions yet.

**Comment by moving-bits** [comment]
> Might be due to [...] power management on newer androud devices.

This might very well be the case, as Android has enforced power management restrictions a lot throughout newer Android versions.

**Comment by moving-bits** [comment]
Setting the "do not merge" label again on this one, as I've implemented a different approach for proximity notification in map's cache and waypoint popups, which uses the same tts navigation as CompassActivity, see PR #7670.
(Proximity "beeps" could be an additional notification as it plays its sounds automatically without further user action necessary, once activated, but maybe #7670 already brings enough functionality for proximity notification and is consistent with compass' behavior. TBD later.)

**Comment by SammysHP** [comment]
I'm not sure I've understood your intention. TTS and proximity beep are two distinct features. While TTS is permanently active and repeats direction and distance, proximity beep is a single event (the first time the distance is low enough after starting navigation). They can be enabled individually.

Btw: The enable status of proximity beep should be persistent.

**Comment by Lineflyer** [comment]
We could use TTS for cache specific proximity notification and a beep for every cache which comes close while live map is running. 
I am just testing this POI alarm with Locus and I have to admit its very nice while on a biking tour to notice, that a cache is on the route.

**Comment by SammysHP** [comment]
Quick example from my Garmin device: https://youtu.be/SkRAnZikhr8

This is what I originally intended with this feature request. There is also a proximity alert that can be configured for all wayoints.

**Comment by Lineflyer** [comment]
In order to keep a bit more overview in this discussion I will continue in #417

**Comment by moving-bits** [comment]
Due to the discussion in #417 PR needs to be rethought. Therefore closing for now.

**Pull request closed without merging**